---
title: "2. Signature analysis with signature-specific cutoffs"
author: "Lea Jopp-Saile and Daniel Hübschmann"
date: "27/11/2019"
vignette: >
  %\VignetteIndexEntry{2. Signature analysis with signature-specific cutoffs}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}  
output: 
  BiocStyle::html_document:
    toc: yes
    
references:
- author:
  - family: Alexandrov
    given: LB
  - family: Nik-Zainal
    given: S
  - family: Wedge
    given: DC
  - family: Aparicio
    given: SA
  - family: Behjati
    given: S
  - family: Biankin
    given: AV
  - family: Bignell
    given: GR
  - family: Bolli
    given: N
  - family: Borg
    given: A
  - family: Borresen-Dale
    given: AL
  - family: Boyault
    given: S
  - family: Burkhardt
    given: B
  - family: Butler
    given: AP
  - family: Caldas
    given: C
  - family: Davies
    given: HR
  - family: Desmedt
    given: C
  - family: Eils
    given: R
  - family: Eyfjörd
    given: JE
  - family: Greaves
    given: M
  - family: Hosoda
    given: F
  - family: Hutter
    given: B
  - family: Ilicic
    given: T
  - family: Imbeaud
    given: S
  - family: Imielinski
    given: M
  - family: Jäger
    given: N
  - family: Jones
    given: DT
  - family: Jones
    given: D
  - family: Knappskog
    given: S
  - family: Kool
    given: M
  - family: Lakhani
    given: SR
  - family: Lopez-Otin
    given: C
  - family: Martin
    given: S
  - family: Munshi
    given: NC
  - family: Nakamura
    given: H
  - family: Northcott
    given: PA
  - family: Pajic
    given: M
  - family: Papaemmanuil
    given: E
  - family: Paradiso
    given: A
  - family: Pearson
    given: JV
  - family: Puente
    given: XS
  - family: Raine
    given: K
  - family: Ramakrishna
    given: M
  - family: Richardson
    given: AL
  - family: Richter
    given: J
  - family: Rosenstiel
    given: P
  - family: Schlesner
    given: M
  - family: Schumacher
    given: TN
  - family: Span
    given: PN
  - family: Teague
    given: JW
  - family: Tokoti
    given: Y
  - family: Tutt
    given: AN
  - family: Valdes-Mas
    given: R
  - family: van Buuren
    given: MM
  - family: van't Veer
    given: L
  - family: Vincent-Salomon
    given: A
  - family: Waddell
    given: N
  - family: Yates
    given: LR
  - family: Australian Pancreatic Cancer Initiative
    given: 
  - family: ICGC Breast Cancer Consortium
    given: 
  - family: ICGC MMML-Seq Consortium
    given: 
  - family: ICGC PedBrain
    given: 
  - family: Zucman-Rossi
    given: J
  - family: Futreal
    given: PA
  - family: McDermott
    given: U
  - family: Lichter
    given: P
  - family: Meyerson
    given: M
  - family: Grimmond
    given: SM
  - family: Siebert
    given: R
  - family: Campo
    given: E
  - family: Shibata
    given: T
  - family: Pfister
    given: SM
  - family: Campbell
    given: PJ
  - family: Stratton
    given: MR
  container-title: Nature
  id: Alex2013
  issued:
    month: 08
    volume: 500
    year: 2013
  publisher: Nature Publishing Group
  title: 'Signatures of Mutational Processes in Cancer'
- author:
  - family: Alexandrov
    given: LB
  - family: Kim
    given: J
  - family: Haradhvala
    given: NJ
  - family: Huang
    given: MN
  - family: Ng
    given: AW
  - family: Boot
    given: A
  - family: Covington
    given: KR
  - family: Gordenin
    given: DA
  - family: Bergstrom
    given: E
  - family: Lopez-Bigas
    given: N
  - family: Klimczak
    given: LJ
  - family: McPherson
    given: JR
  - family: Morganella
    given: S
  - family: Sabarinathan
    given: R
  - family: Wheeler
    given: DA
  - family: Mustonen
    given: V
  - family: Getz
    given: G
  - family: Rozen
    given: SG
  - family: Stratton
    given: MR
  container-title: bioarxiv
  id: Alex2018
  issued:
    month: 05
    year: 2018
  publisher: bioarxiv
  title: 'The Repertoire of Mutational Signatures in Human Cancer'
- author:
  - family: Alexandrov
    given: LB
  - family: Nik-Zainal
    given: S
  - family: Wedge
    given: DC
  - family: Campbell
    given: PJ
  - family: Stratton
    given: MR
  container-title: Cell Reports
  id: Alex_CellRep2013
  issued:
    year: 2013
  title: > 
    Deciphering signatures of mutational processes operative in human cancer
- author:
  - family: Wang
    given: Yong
  - family: Lawson
    given: Charles L.
  - family: Hanson
    given: Richard J
  container-title: R-package version 1.1-1
  id: lsei_package2015
  issued:
    year: 2015
  title: >
    lsei: Solving Least Squares Problems under Equality/Inequality Constraints
- author:
  - family: Sing
    given: Tobias
  - family: Sander
    given: Oliver
  - family: Beerenwinkel
    given: Niko
  - family: Lengauer
    given: Thomas
  container-title: Bioinfomatics
  id: ROCR_package2005
  issued:
      year: 2005
  title: >
    ROCR: visualizing classifier performance in R
---

```{r load_style, warning=FALSE, echo=FALSE, message=FALSE, results="hide"}
library(BiocStyle)
```

```{r packages, include=FALSE}
library(YAPSA)
library(knitr)
library(gridExtra)
library(dplyr)
opts_chunk$set(echo=TRUE)
opts_chunk$set(fig.show='asis')
```

# Introduction {#introduction}

The vignette `1. Usage of YAPSA` describes the general usage of YAPSA as a tool
to perform supervised mutational signature analysis. Calling of mutational
signature in YAPSA is performs with a Non-Negative Least Square (NNLS) algorithm
using  **l**inear **c**ombination **d**ecomposition (`LCD`) function based on
the function `lsei` from the package `r CRANpkg("lsei")`[@lsei_package2015]. The
LCD algorithm is highly sensitive and specific resulting in a large number of
false-positive called signatures. The performance of the algorithm can be
improved with trained signature-specific cutoffs, reducing the number of falsely
called signatures. 

LCD determines the contribution of signature in a three-step computational
process 

1. Find an NNLS solution with all supplied signatures using LCD
1. Identify those signatures which have a contribution higher than a
signature-specific cutoff
1. Rerun NNLS based on the selected subset of signatures

Mutational signatures stem from an NMF analysis and their signals are not
orthogonal. This causes overlapping of the signature signal and a difference in
detectability. These two aspects are taken into account by the training process
of signature-specific cutoffs. The training of signature-specific cutoffs was
performed before-hand and values stored as a dataframes alongside with the
numerical values of the contributions of the different features to the
signatures. 


The signature-specific cutoffs integrated into YAPSA were trained by an
optimizing procedure using a modified Receiver Operating Characteristic (ROC)
[@ROCR_package2005]. Hereby, an NNLS algorithm using LCD was trained to yield a
maximum overlap with de novo NMF analysis by Alexandrov et al. on a cross-entity
collection of cancer samples. A cost function of the ROC analysis was defined by
cost terms punishing false-negative and false-positive findings separately. The
ratio between the cost for a false-negative divided by the cost for a false
positive was called the cost factor
(cost factor=(False-positive findings)/(False-negative findings)).
Individual ROC analyses were performed for every signature. The global minimum
of the cost function defines the signature-specific cutoff value. The shape of
the function and hence its minimum is depended on the cost factor. Through an
optimization procedure using varying punishing costs for false-positive findings
and constant punishing cost of one for false-negative findings, the minimum of
the relative false attribution was determined. Based on the cost factor the
optimal cutoff values were computed. 


# Data overview

In total YAPSA comes with 15 signature dataframes, which can be loaded as
followed (more information provided by the help function):
```{r load signatures}
data(sigs)
data(sigs_pcawg)
```

The command `data(sigs)` will load eight signatures dataframes, whereas four
dataframes containing the mutational signatures themselves and the remaining
four data contain additional information on the signatures. The eight signature
dataframes are comprised of:

* Two signature and two additonal information dataframes with SNV Alexandrov
COSMIC signatures (<https://cancer.sanger.ac.uk/cosmic/signatures_v2>),
downloaded on January 15th, 2016): `AlexCosmicArtif_sig_df`,
`AlexCosmicValid_sig_df`, `AlexCosmicArtif_sigInd_df` and
`AlexCosmicValid_sigInd_df`

* Two signature and two additonal information dataframes dataframes with SNV
Alexadrov Initial signatures (as available at ???): `AlexInitialArtif_sig_df`,
`AlexInitialValid_sig_df`, `AlexInitialArtif_sigInd_df` and
`AlexInitialValid_sigInd_df`

The command `data(sigs_pcawg)` will load six signatures dataframes, whereas
three dataframes containing the mutational signatures themselves and the
remaining three data contain additional information on the signatures. The six
signature dataframes are comprised of:

* Two signature and two additional information dataframes with SNV Alexandrov
PCAWG signatures (<https://cancer.sanger.ac.uk/cosmic/signatures_v3>),
downloaded on August 16th, 2018): `PCAWG_SP_SBS_sigs_Artif_df`,
`PCAWG_SP_SBS_sigs_Real_df`, `PCAWG_SP_SBS_sigInd_Artif_df` and
`PCAWG_SP_SBS_sigInd_Real_df`

* One signature and oen additional information dataframe with Indel Alexandrov
PCAWG signature (<https://cancer.sanger.ac.uk/cosmic/signatures_v3>), downloaded
on August 16th, 2018): `PCAWG_SP_ID_sigs_df`, `PCAWG_SP_ID_sigInd_df`

For each of the signature dataframe at least one corresponding
signature-specific cutoff dataframe is available. Cutoff dataframes can be
loaded equivalent to the signatures
```{r load cutoffs}
data(cutoffs)
data(cutoffs_pcawg)
```

The command `data(cutoffs)` will load for the four COSMIC and initial Alexandrov
signature frames two corresponding cutoff dataframes, e.g. For
`AlexCosmicValid_sig_df` the corresponding cutoff dataframes are
`cutoffCOSMICValid_abs_df` and `cutoffCOSMICValid_rel_df`. Absolute (abs)
signature-specific cutoffs should be used for the analysis of whole genome
sequencing (WGS) data, whereas relative (rel) signature-specific cutoffs are
valid for the analysis of whole exome sequencing (WES) data. The two different
cutoffs account for the different maximal feature occurrence between WGS and WES
data.  

With `data(cutoffs_pcawg)` the signature-specific cutoff dataframes for the
PCAWG signatures are loaded. In this case, only one dataframe with cutoffs per
signature dataframe is available. Signature-specific cutoffs are valid for WGS
and WES data analysis as differences in the feature occurrence is taken into
account during the training procedure.

Each dataframes contains several rows, corresponding to the different cost
factors. The cost factor which determines the optimal cutoff value is listed in
the table below. 

cutoff dataframe   | cost factor
------- | -----
cutoffInitialValid_abs_df | ?
cutoffInitialValid_rel_df | ?
cutoffInitialArtif_abs_df | ?
cutoffInitialArtif_rel_df | ?
cutoffCosmicValid_abs_df | 6
cutoffCosmicValid_rel_df | 6
cutoffCosmicArtif_abs_df | 6
cutoffCosmicArtif_rel_df | 6
cutoffPCAWG_SBS_WGSWES_realPid_df | 10
cutoffPCAWG_SBS_WGSWES_artiflPid_df | 16
cutoffPCAWG_ID_WGS_Pid_df | 3

: (\#tab:Table:) Overview of signature specific cutoffs and the cost factor to
obtain the optimal cutoffs.

# Example
The `1. Usage of YAPSA` shows how to obtain a mutational catalog ($V$) from a
vcf file, which is in the following loaded from the provided data. More details
on the data set are provided in `1. The usage of YAPSA` vignette.

```{r load data}
#load("/Users/jopp-saile/software/yapsa-bioconductor/data/lymphomaNature2013_mutCat_df.rda")
data(lymphomaNature2013_mutCat_df)
```

The mutational catalog $V$ together with the signatures $V$ and the
corresponding signature-specific cutoffs can be used to determine the signature
contribution ($H$) per sample or cohort. 
The function `LCD_complex_cutoff_combined` allows computing the signature
contribution per patient identifier (PID), sample and the consensus of both
analysis respectively.

In `1. The usage of YAPSA` signature contribution was determined without a
signature-specific cutoff (zero cutoff vector). In the following, the aim is to
identify the significant contribution of validated COSMIC signatures.

```{r set variables }
current_sig_df <- AlexCosmicValid_sig_df
current_sigInd_df <- AlexCosmicValid_sigInd_df
```

Creating a zero cutoff vector:
```{r cutoff vector}
current_cutoff_vector <- rep(0, dim(AlexCosmicValid_sig_df)[2])
```

Identification of the signature contiburion of validaded COSMIC signatures with
a zero cutoff vector using LCD analysis.
```{r lymphoma_cohort_LCD_results}
lymphoma_COSMICExposures_zero_df <-
  LCD_complex_cutoff_combined(
    in_mutation_catalogue_df = lymphomaNature2013_mutCat_df,
    in_cutoff_vector = current_cutoff_vector, 
    in_signatures_df = current_sig_df, 
    in_sig_ind_df = current_sigInd_df)

```
The function returns an object with the results per cohort, per PID and the
consensus between both analyses.

Annotation of a subgroup to each PID allows to them per subgroup
```{r subrgroup annotation}
data(lymphoma_PID)
colnames(lymphoma_PID_df) <- "SUBGROUP"
lymphoma_PID_df$PID <- rownames(lymphoma_PID_df)
COSMIC_subgroups_df <- 
  make_subgroups_df(lymphoma_PID_df,
                    lymphoma_COSMICExposures_zero_df$cohort$exposures)
```

```{r caption_barplot_2, echo=FALSE}
cap <- "Absoute exposures of the COSMIC signatures in the lymphoma mutational
        catalogs, signature-specific cutoffs with a cost factor of 6 used
        for the LCD"
```

Exposures of a cohort-wide analysis identified with LCD and a zero-cutoff vector
are represented in a stacked barplot per patient. Patients have a small
contribution to multiple signatures.
```{r exposures_zero, warning=FALSE, fig.width=8, fig.height=6, fig.cap= cap}
result_cohort <- lymphoma_COSMICExposures_zero_df$cohort
exposures_barplot(
  in_exposures_df = result_cohort$exposures,
  in_signatures_ind_df = result_cohort$out_sig_ind_df,
  in_subgroups_df = COSMIC_subgroups_df)        
```

The zero-cutoff vector will now be replaced with the signature-specific cutoff
vector. After the initial LCD analysis only signatures that contribute higher
then their signature specific cutoff value will be considered, causing a
reduction of the overall number of detected signatures.


Based on \@ref(tab:Table1) we know that for the validated COSMIC Signatures the
optimal cutoffs are the once computed with a cost factor of six. 
```{r set signature-specific cutoff}
current_cutoff_df <- cutoffCosmicValid_abs_df
current_cost_factor <- 6
current_cutoff_vector <- current_cutoff_df[current_cost_factor,]
```

Identification of the signature contribution of validated COSMIC signatures with
a zero cutoff vector using LCD analysis.
```{r LCD with cutoffs}
lymphoma_COSMICExposures_cutoffs_df <-
  LCD_complex_cutoff_combined(
      in_mutation_catalogue_df = lymphomaNature2013_mutCat_df,
      in_cutoff_vector = current_cutoff_vector, 
      in_signatures_df = current_sig_df, 
      in_sig_ind_df = current_sigInd_df)
```

Exposures of a cohort-wide analysis identefied with LCD and signature-specific
cutoffs are represented in a stacked barplot per patient. Patients have small
contibution to multiple signatures.
```{r caption_barplot, echo=FALSE}
cap <- "Absolute exposures of the COSMIC signatures in the lymphoma mutational
        catalogs, signature-specific cutoffs with a cost factor of 6 used
        for the LCD"
```

```{r exposures_cutoffs, warning=FALSE, fig.width=8, fig.height=6, fig.cap= cap}
result_cohort <- lymphoma_COSMICExposures_cutoffs_df$cohort
exposures_barplot(
  in_exposures_df = result_cohort$exposures,
  in_signatures_ind_df = result_cohort$out_sig_ind_df,
  in_subgroups_df = COSMIC_subgroups_df)               
```

It can be clearly shown that the number of identified signatures are less, as
the number of false-positive detected signatures are reduced. 

# Reference